pkg/machines/components/create-vm-dialog/pxe-helpers.js

```diff
@@ -0,0 +1,187 @@
+ * @param {string} connectionName - The libvirt connection name.
+ */
+export function getVirtualNetworkByName(virtualNetworkName, virtualNetworks, connectionName) {
+    return virtualNetworks.filter(virtualNetwork =>
+        virtualNetwork.connectionName == connectionName &&
+        virtualNetwork.name == virtualNetworkName
+    )[0];
+}
+
+/**
+ * Return a short description of the Virtual Network.
+ * @param {object} virtualNetwork - A Virtual Network object.
+ */
+function getVirtualNetworkDescription(virtualNetwork) {
+    let mode, dev;
+    let forward = virtualNetwork.forward;
+
+    if (forward) {
+        mode = forward.mode;
+        dev = forward.interface && forward.interface.dev;
+    }
+
+    if (mode || dev) {
+        if (!mode || mode == 'nat') {
+            if (dev)
+                return cockpit.format('NAT to $0', dev);
+            else
+                return 'NAT';
+        } else if (mode == 'route') {
+            if (dev)
+                return cockpit.format('Route to $0', dev);
+            else
+                return 'Routed Network';
+        } else {
+            if (dev)
+                return cockpit.format('$0 to $1', mode, dev);
+            else
+                return cockpit.format('$0 network', mode.toUpperCase());
+        }
+    } else {
+        return _("Isolated network");
+    }
+}
+
+/**
+ * @param {array} nodeDevices - An array of object containing NodeDevices.
+ * @param {object} virtualNetwork - A Virtual Network object.
+ */
+export function getVirtualNetworkPXESupport(virtualNetwork) {
+    if (virtualNetwork.forward && virtualNetwork.forward.mode != 'nat') {
+        return true;
+    }
+
+    let ips = virtualNetwork.ip;
+
+    for (let i = 0; i < ips.length; i++)
+        if (ips[i].dhcp.bootp)
+            return true;
+
+    return false;
+}
+
+/**
+ * Returns the first available Network Resource to be used for showing to PXE Network Sources list.
+ * @param {array} nodeDevices - An array of object containing NodeDevices.
+ * @param {array} virtualNetworks - An array of object containing Virtual Networks.
+ * @param {string} connectionName - The libvirt connection name.
+ */
+export function getPXEInitialNetworkSource(nodeDevices, virtualNetworks, connectionName) {
+    if (virtualNetworks.length > 0)
+        return cockpit.format('network=$0', virtualNetworks[0].name);
+
+    let netNodeDevices = getNodeDevicesOfType(nodeDevices, connectionName, 'net');
+
+    /* Do not show to the user the libvirt virbrX-nic devices since these are
+     * supposed to be managed through Virtual Networks
+     */
+    let libvirtVirBridges = getLibvirtNetworkBridges(virtualNetworks);
+    netNodeDevices = netNodeDevices.filter(netNodeDevice => {
+        for (let i in libvirtVirBridges) {
+            if (netNodeDevice.capability.interface == (libvirtVirBridges[i] + '-nic'))
+                return false;
+        }
```

Comment:
This is a quadratic loop again, let's just hope that users don't have too many devices and bridges :-)

This could be optimized a bit by checking in advance whether netNodeDevice's interface ends with `-nic`.

Reply:
Optimization mentioned in second line done. For time complexity I think we should not worry too much even for 100 net devices it will get a O(n3) -> 100^3 which is still taking less than a second. I tend to count approximately 10^7 operations as 1 second.